[Feature][jdbc] 优化ClassCastException， 优化写入时数据库连接无效问题

### Search before asking

- [X] I had searched in the [issues](https://github.com/DTStack/chunjun/issues) and found no similar feature requirement.


### Description


1、修复bigint、int等无符号类型的转换失败问题

```

2023-04-17 15:07:07,637 ERROR com.dtstack.chunjun.source.DtInputFormatSourceFunction       [] - Exception happened, start to close format
org.apache.flink.streaming.runtime.tasks.ExceptionInChainedOperatorException: Could not forward element to next operator
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:96) ~[flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:57) ~[flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:29) ~[flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:56) ~[flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:29) ~[flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$ManualWatermarkContext.processAndCollect(StreamSourceContexts.java:418) ~[flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$WatermarkContext.collect(StreamSourceContexts.java:513) ~[flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$SwitchingOnClose.collect(StreamSourceContexts.java:103) ~[flink-dist-1.16.1.jar:1.16.1]
	at com.dtstack.chunjun.source.DtInputFormatSourceFunction.run(DtInputFormatSourceFunction.java:127) [chunjun-core.jar:?]
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:110) [flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:67) [flink-dist-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:333) [flink-dist-1.16.1.jar:1.16.1]
Caused by: java.lang.ClassCastException: java.math.BigInteger cannot be cast to java.lang.Long
	at org.apache.flink.table.data.GenericRowData.getLong(GenericRowData.java:154) ~[flink-table-api-java-uber-1.16.1.jar:1.16.1]
	at org.apache.flink.table.data.RowData.lambda$createFieldGetter$245ca7d1$7(RowData.java:249) ~[flink-table-api-java-uber-1.16.1.jar:1.16.1]
	at org.apache.flink.table.data.RowData.lambda$createFieldGetter$25774257$1(RowData.java:296) ~[flink-table-api-java-uber-1.16.1.jar:1.16.1]
	at org.apache.flink.table.runtime.typeutils.RowDataSerializer.copyRowData(RowDataSerializer.java:170) ~[flink-table-runtime-1.16.1.jar:1.16.1]
	at org.apache.flink.table.runtime.typeutils.RowDataSerializer.copy(RowDataSerializer.java:131) ~[flink-table-runtime-1.16.1.jar:1.16.1]
	at org.apache.flink.table.runtime.typeutils.RowDataSerializer.copy(RowDataSerializer.java:48) ~[flink-table-runtime-1.16.1.jar:1.16.1]
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:80) ~[flink-dist-1.16.1.jar:1.16.1]
	... 11 more
2023-04-17 15:07:09,639 INFO  com.dtstack.chunjun.source.format.BaseRichInputFormat        [] - subtask input close finished
2023-04-17 15:07:09,640 INFO  com.dtstack.chunjun.sink.format.BaseRichOutputFormat         [] - taskNumber[1] close()
2023-04-17 15:07:11,642 INFO  com.dtstack.chunjun.sink.format.BaseRichOutputFormat         [] - subtask[1}] close() finished
2023-04-17 15:07:11,643 WARN  org.apache.flink.runtime.taskmanager.Task                    [] - Source: source[1] -> Calc[2] -> Sink: sink[3] (2/2)#0 (56f369d84a2adb5f6b6251fe0728b569_cbc357ccb763df2852fee8c4fc7d55f2_1_0) switched from RUNNING to FAILED with failure cause: java.lang.ClassCastException: java.math.BigInteger cannot be cast to java.lang.Long
	at org.apache.flink.table.data.GenericRowData.getLong(GenericRowData.java:154)
	at org.apache.flink.table.data.RowData.lambda$createFieldGetter$245ca7d1$7(RowData.java:249)
	at org.apache.flink.table.data.RowData.lambda$createFieldGetter$25774257$1(RowData.java:296)
	at org.apache.flink.table.runtime.typeutils.RowDataSerializer.copyRowData(RowDataSerializer.java:170)
	at org.apache.flink.table.runtime.typeutils.RowDataSerializer.copy(RowDataSerializer.java:131)
	at org.apache.flink.table.runtime.typeutils.RowDataSerializer.copy(RowDataSerializer.java:48)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.pushToOperator(CopyingChainingOutput.java:80)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:57)
	at org.apache.flink.streaming.runtime.tasks.CopyingChainingOutput.collect(CopyingChainingOutput.java:29)
	at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:56)
	at org.apache.flink.streaming.api.operators.CountingOutput.collect(CountingOutput.java:29)
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$ManualWatermarkContext.processAndCollect(StreamSourceContexts.java:418)
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$WatermarkContext.collect(StreamSourceContexts.java:513)
	at org.apache.flink.streaming.api.operators.StreamSourceContexts$SwitchingOnClose.collect(StreamSourceContexts.java:103)
	at com.dtstack.chunjun.source.DtInputFormatSourceFunction.run(DtInputFormatSourceFunction.java:127)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:110)
	at org.apache.flink.streaming.api.operators.StreamSource.run(StreamSource.java:67)
	at org.apache.flink.streaming.runtime.tasks.SourceStreamTask$LegacySourceFunctionThread.run(SourceStreamTask.java:333)

```


2、优化 写入时数据库连接无效问题

当写入mysql时，mysql服务端的  wait_timeout 设置为10 秒过短时，读取数据超过 10 秒后再写入时会报错数据库连接 丢失

![image](https://user-images.githubusercontent.com/3014381/232689939-81b9ce04-bc81-4a33-895f-6d1a3dbd46f0.png)





### Use case



**第一个问题，测试用例：**


mysql 表结构：

```
CREATE TABLE flinkx_test.dc_user (
  id bigint(20) UNSIGNED NOT NULL AUTO_INCREMENT,
  username varchar(32) NOT NULL DEFAULT 'nothing',
  PRIMARY KEY (id)
)
ENGINE = INNODB,
AUTO_INCREMENT = 8397,
AVG_ROW_LENGTH = 223,
CHARACTER SET utf8,
COLLATE utf8_general_ci,
COMMENT = '用户信息表',
ROW_FORMAT = COMPACT;
```

表里数据：

```
INSERT INTO dc_user (id, username) VALUES
(2, 'test');

```

```

SET chunjun.delay_period_mill=2000;
CREATE TABLE source
(
    `id` bigint
) WITH (
      'password' = '123456',
      'connector' = 'mysql-x',
      'scan.query-timeout' = '600',
      'table-name' = 'dc_user',
      'url' = 'jdbc:mysql://ip:port/flinkx_test?useUnicode=true&characterEncoding=utf-8&useSSL=false&connectTimeout=3000&zeroDateTimeBehavior=convertToNull&useunicode=true&characterEncoding=utf8&useSSL=false&useCursorFetch=true&tinyInt1isBit=false&zeroDateTimeBehavior=convertToNull',
      'username' = 'root'
      );
CREATE TABLE sink
(
    `id` bigint
) WITH (
      'connector' = 'stream-x',
      'print' = 'true'
      );
insert into sink
select
    `id` as `id`
from source;


```

第二个问题测试请自行模拟

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/converter/JdbcSqlConverter.java
+++ b/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/converter/JdbcSqlConverter.java
@@ -112,6 +112,14 @@ public class JdbcSqlConverter
         GenericRowData genericRowData = new GenericRowData(rowType.getFieldCount());
         for (int pos = 0; pos < rowType.getFieldCount(); pos++) {
             Object field = resultSet.getObject(pos + 1);
+            if (resultSet.getMetaData().getColumnTypeName(pos + 1).equals("BIGINT UNSIGNED")
+                    && field != null) {
+                field = ((BigInteger) field).longValue();
+            }
+            if (resultSet.getMetaData().getColumnTypeName(pos + 1).equals("INT UNSIGNED")
+                    && field != null) {
+                field = ((Long) field).intValue();
+            }
             genericRowData.setField(pos, toInternalConverters.get(pos).deserialize(field));
         }
         return genericRowData;

--- a/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/sink/JdbcOutputFormat.java
+++ b/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/sink/JdbcOutputFormat.java
@@ -124,6 +124,11 @@ public class JdbcOutputFormat extends BaseRichOutputFormat {
     protected void writeSingleRecordInternal(RowData row) throws WriteRecordException {
         int index = 0;
         try {
+            if (dbConn.isClosed() || !dbConn.isValid(10)) {
+                dbConn = getConnection();
+                buildStmtProxy();
+                log.warn("write single record reconnecting");
+            }
             stmtProxy.writeSingleRecordInternal(row);
             if (Semantic.EXACTLY_ONCE == semantic) {
                 rowsOfCurrentTransaction += rows.size();
@@ -150,6 +155,19 @@ public class JdbcOutputFormat extends BaseRichOutputFormat {
     @Override
     protected void writeMultipleRecordsInternal() throws Exception {
         try {
+            // 解决写入时数据库连接无效问题
+            try {
+                if (!dbConn.isValid(10)) {
+                    dbConn = getConnection();
+                    buildStmtProxy();
+                    log.warn("write multiple records reconnecting");
+                }
+            } catch (Throwable e) {
+                // 解决 sqlserver JtdsConnection 没有实现 isValid 方法
+                if (!(e instanceof AbstractMethodError)) {
+                    e.printStackTrace();
+                }
+            }
             for (RowData row : rows) {
                 stmtProxy.convertToExternal(row);
                 stmtProxy.addBatch();

--- a/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/source/JdbcInputFormat.java
+++ b/chunjun-connectors/chunjun-connector-jdbc-base/src/main/java/com/dtstack/chunjun/connector/jdbc/source/JdbcInputFormat.java
@@ -216,6 +216,7 @@ public class JdbcInputFormat extends BaseRichInputFormat {
             }
             return finalRowData;
         } catch (Exception se) {
+            log.error(ExceptionUtil.getErrorMessage(se));
             throw new ReadRecordException("", se, 0, rowData);
         } finally {
             try {

--- a/chunjun-connectors/chunjun-connector-mysql/src/main/java/com/dtstack/chunjun/connector/mysql/converter/MysqlRawTypeConverter.java
+++ b/chunjun-connectors/chunjun-connector-mysql/src/main/java/com/dtstack/chunjun/connector/mysql/converter/MysqlRawTypeConverter.java
@@ -32,6 +32,10 @@ public class MysqlRawTypeConverter {
      * com.mysql.jdbc.ResultSetImpl.getObject(int)
      */
     public static DataType apply(String type) {
+        // 修复不识别带有小括号精度类型问题
+        if (type.indexOf("(") > -1) {
+            type = type.substring(0, type.indexOf("("));
+        }
         switch (type.toUpperCase(Locale.ENGLISH)) {
             case "BOOLEAN":
             case "BIT":

